specs/lazy-adr/adr-010-exec-api.md

```diff
@@ -0,0 +1,274 @@
+
+## Changelog
+
+- 2025.01.13: Initial draft
+
+## Context
+
+Introduction of the Execution API makes rollkit very generic and execution-environment agnostic.
+It removes all ABCI-centric code for full interoperability with other types of VMs.
+
+## Alternative Approaches
+
+1. Maintain current state: keep ABCI interface and implement other VMs inside ABCI application.
+2. Migrate to Engine API.
+3. Create new API generic enough to handle any arbitrary VM.
+
+## Decision
+
+New generic Execution API is proposed.
+It was designed to accommodate ABCI, Engine API, and any other VM / execution environment.
+
+## Detailed Design
+
+### API
+
+Execution API consist of 4 methods defined in [go-execution](https://github.com/rollkit/go-execution) repository.
+
+```go
+// Executor defines a common interface for interacting with the execution client.
+type Executor interface {
+	// InitChain initializes the blockchain with genesis information.
+	InitChain(ctx context.Context, genesisTime time.Time, initialHeight uint64, chainID string) (stateRoot types.Hash, maxBytes uint64, err error)
+
+	// GetTxs retrieves all available transactions from the execution client's mempool.
+	GetTxs(ctx context.Context) ([]types.Tx, error)
+
+	// ExecuteTxs executes a set of transactions to produce a new block header.
+	ExecuteTxs(ctx context.Context, txs []types.Tx, blockHeight uint64, timestamp time.Time, prevStateRoot types.Hash) (updatedStateRoot types.Hash, maxBytes uint64, err error)
+
+	// SetFinal marks a block at the given height as final.
+	SetFinal(ctx context.Context, blockHeight uint64) error
+}
+```
+
+### API Methods specification
+
+#### `InitChain`
+
+##### Description
+
+Initializes the blockchain's state based on genesis information. This method is invoked at the beginning of the blockchain's lifecycle to prepare the execution environment for subsequent operations.
+
+##### Inputs
+
+- `ctx` (`context.Context`): Context for managing request timeouts and cancellations.
+- `genesisTime` (`time.Time`): The initial timestamp of the rollup.
+- `initialHeight` (`uint64`): The starting height of the rollup.
+- `chainID` (`string`): A unique identifier of the rollup network.
+
+##### Outputs
+
+- `stateRoot` (`types.Hash`): The resulting state root after initializing the chain.
+- `maxBytes` (`uint64`): Maximum block size in bytes, as defined by the execution client's genesis configuration.
+- `err` (`error`): An error, if the initialization process fails.
+
+##### Expected Behavior
+
+- Initialize rollup according to the genesis.
+- Generate an initial `stateRoot` representing the genesis state of the rollup.
+- Return the maximum allowable block size (`maxBytes`).
+
+#### `GetTxs`
+
+##### Description
+
+Fetches all pending transactions from the execution client's mempool.
+Transactions returned by execution client will be passed by rollkit to sequencer.
+
+##### Inputs
+
+- `ctx` (`context.Context`): Context for managing request timeouts and cancellations.
```

Comment:
where is the block created and how does the app know how many txs to return? does it return the whole mempool?

Reply:
Entire mempool contents should be returned. Sequencer should peek the best transactions that fits in single block (`maxBytes`) and send them back. Transactions included in block must be removed from mempool.